lib/http3/qpack.c

```diff
@@ -783,13 +785,32 @@ static int parse_decode_context(h2o_qpack_decoder_t *qpack, struct st_h2o_qpack_
             return H2O_HTTP3_ERROR_QPACK_DECOMPRESSION_FAILED;
     }
 
-    /* base index */
+    /* sign and delta base */
     if (*src >= src_end)
         return H2O_HTTP3_ERROR_QPACK_DECOMPRESSION_FAILED;
     int sign = (**src & 0x80) != 0;
-    if (decode_int(&ctx->base_index, src, src_end, 7) != 0)
+    int64_t delta_base;
+    if (decode_int(&delta_base, src, src_end, 7) != 0)
         return H2O_HTTP3_ERROR_QPACK_DECOMPRESSION_FAILED;
-    ctx->base_index = sign == 0 ? ctx->req_insert_count + ctx->base_index : ctx->req_insert_count - ctx->base_index - 1;
+
+    assert(0 <= ctx->req_insert_count && ctx->req_insert_count <= INT64_MAX);
+    assert(0 <= delta_base && delta_base <= INT64_MAX);
+
+    /* base index */
+    if (sign == 0) {
+        if (delta_base > INT64_MAX - ctx->req_insert_count) {
+            return H2O_HTTP3_ERROR_QPACK_DECOMPRESSION_FAILED;
+        }
+        ctx->base_index = ctx->req_insert_count + delta_base;
+    } else {
+        /* we don't accept negative base index though current QPACK specification doesn't mention to such case */
+        /* let's keep our eyes on https://github.com/quicwg/base-drafts/issues/4938 */
+        if (ctx->req_insert_count - 1 < delta_base) {
+            return H2O_HTTP3_ERROR_QPACK_DECOMPRESSION_FAILED;
+        }
+        ctx->base_index = ctx->req_insert_count - delta_base - 1;
+    }
+    assert(0 <= ctx->base_index && ctx->base_index <= INT64_MAX);
```

Comment:
I think the upper bound should be smaller than INT64_MAX, because we do in `resolve_dynamic_postbase` we do `base_index + 1`.

How about changing the valid maximum value of `req_insert_count` and `delta_base` from 1<sup>63</sup> - 1 to 1<sup>62</sup> - 1? The rationale is that a QPACK encoder using a QUIC v1 stream can never create more than 1<<62 QPACK table entries. That is because the smallest size of a QPACK encoder instruction is 1 byte, and because the length of a QPACK encoder stream can be no longer than 1<sup>62</sup> bytes.

If we make such a change, then we can be certain that `base_index` would never be as large as 1<sup>63</sup> - 1.